app/services/info_request_batch_zip.rb

```diff
@@ -0,0 +1,98 @@
+##
+# This service streams a ZIP file with all incoming/ outgoing messages and
+# attachments for each request for a given info request batch
+#
+class InfoRequestBatchZip
+  include DownloadHelper
+
+  ZippableFile = Struct.new(:path, :body)
+
+  attr_reader :info_request_batch
+
+  def initialize(info_request_batch)
+    @info_request_batch = info_request_batch
+  end
+
+  def files
+    files = [prepare_dashboard_metrics]
+
+    info_request_events.each_with_object(files) do |event, memo|
+      if event.outgoing?
+        memo << perpare_outgoing_message(event.outgoing_message)
+      elsif event.response?
+        memo << perpare_incoming_message(event.incoming_message)
+        event.incoming_message.foi_attachments.each do |attachment|
+          memo << perpare_foi_attachment(attachment)
+        end
+      end
+    end
+  end
+
+  def name
+    generate_download_filename(
+      resource: 'batch',
+      id: info_request_batch.id,
+      title: info_request_batch.title,
+      type: 'export',
+      ext: 'zip'
+    )
+  end
+
+  def stream(&chunks)
+    block_writer = ZipTricks::BlockWrite.new(&chunks)
+
+    ZipTricks::Streamer.open(block_writer) do |zip|
+      files.each do |file|
+        zip.write_deflated_file(file.path) { |writer| writer << file.body }
```

Comment:
At the moment, this isn't initiating the download until most of the work is done. `files` is doing all of the database querying and only _then_ do we start to stream each file. On a large batch this means a ~30 second wait until we get an indication that something is happening.

You can see this happening in the logs with:

```diff
 files.each do |file|
+  ::Rails.logger.info "WRITING FILE: #{file.path}"
   zip.write_deflated_file(file.path) { |writer| writer << file.body }
 end
```

Instead, we want to initiate the download straight away. Here's an example:

```ruby
  ZipTricks::Streamer.open(block_writer) do |zip|
      (1..100).each do |i|
         zip.write_deflated_file(i.to_s) do |file_writer|
           ::Rails.logger.info "WRITING FILE: #{i}"
           file_writer << (i.to_s * 1_000_000)
         end
      end
  end
```

This will illustrate the download initiating, and then writing a file at a time as the Zip continues to download.

For our app code, I think we want something more like this. **WARNING** this does start the download more immediately, but for some reason it doesn't write incoming messages or attachments to the Zip, so take care to test that it works.

```ruby
    ZipTricks::Streamer.open(block_writer) do |zip|
      file = prepare_dashboard_metrics
      zip.write_deflated_file(file.path) { |writer| writer << file.body }

      info_request_events.each do |event|
        if event.outgoing?
          file = perpare_outgoing_message(event.outgoing_message)
          ::Rails.logger.info "WRITING FILE: #{file.path}"
          zip.write_deflated_file(file.path) { |writer| writer << file.body }
        elsif event.response?
          file = perpare_incoming_message(event.incoming_message)
          ::Rails.logger.info "WRITING FILE: #{file.path}"
          zip.write_deflated_file(file.path) { |writer| writer << file.body }
          event.incoming_message.foi_attachments.each do |attachment|
            file = perpare_foi_attachment(attachment)
            ::Rails.logger.info "WRITING FILE: #{file.path}"
            zip.write_deflated_file(file.path) { |writer| writer << file.body }
          end
        end
      end
```

Reply:
Resolved by 06b418a